### PR TITLE
i#5694 core-sharded: Count threads in basic_counts

### PR DESCRIPTION
**Pull request by derekbruening**
Adds first-class counting of threads per shard in the basic_counts tool, for use with core-sharded operation.

Updates the core-sharded tests with sanity checks.

Issue: #5694

### CODE REVIEW

File: clients/drcachesim/tools/basic_counts.cpp [hunk 1]
```diff
@@ -282,8 +286,10 @@ basic_counts_t::print_counters(const counters_t &counters, int64_t num_threads,
               << " icache flushes\n";
     std::cerr << std::setw(12) << counters.dcache_flushes << prefix
               << " dcache flushes\n";
-    if (num_threads > 0) {
-        std::cerr << std::setw(12) << num_threads << prefix << " threads\n";
+    if (shard_type_ != SHARD_BY_THREAD || counters.unique_threads.size() > 1 ||
+        prefix == " total") {
```
Comment (abhinav92003) [inline clients/drcachesim/tools/basic_counts.cpp:290]: Named constant would be better.

File: clients/drcachesim/tools/basic_counts.h [hunk 2]
```diff
@@ -119,6 +119,9 @@ class basic_counts_t : public analysis_tool_t {
                     unique_pc_addrs.insert(addr);
                 }
             }
+            for (const memref_tid_t tid : rhs.unique_threads) {
+                unique_threads.insert(tid);
```
Comment (abhinav92003) [inline clients/drcachesim/tools/basic_counts.h:123]: Could you use range insertion?

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/tools/basic_counts.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/tools/basic_counts.cpp:290]: Done.

File: clients/drcachesim/tools/basic_counts.h (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/tools/basic_counts.h:123]: Yes, done, and for the one above.

**Comment by derekbruening** [comment]
x64 ubuntu22 failure is api.rseq synchronize failure #6185

**Pull request merged**